### ISSUE TITLE
[question] pre-calculated histogram of hdf5 image vs histogram of other image format

### ISSUE THREAD

**Issue by kswang1029** [CARTAvis/carta-backend#235]
By comparing the clip values of an hdf5 image and its counterpart formats (e.g. fits), it appears that the histograms of the two are slightly different. Are they calculated in different ways (hdf5 converter vs carta backend)?

example:
supermosaic.10.hdf5 with clip 99.9% (pre-calculated histogram at 0th channel)
clip min = -26.95503284264798
clip max = 26.57820493466349

supermosaic.10.fits with clip 99.9% (backend-calculated histogram at 0th channel)
clip min = -26.95912703791012
clip max = 26.573279209750027

**Comment by veggiesaurus** [comment]
possibly a rounding issue when calculating the histograms. Do we know which is "right"?

**Comment by kswang1029** [comment]
first thing to check, if clip=100%, is clip min equal to data min? Because it is histogram, do we set clip min to the center of the first bin or the left boundary of the first bin? Same question to clip max. I feel the issue is here based on the little difference in the above example, maybe. Rounding issue could be too.

**Comment by veggiesaurus** [comment]
the actual clip values are calculated by the frontend, _from_ the histogram, so that should be the same. 
Are there any differences when clip=100%? I suspect the issue might be how we're binning things that lie very close to the edge of a bin boundary.

**Comment by kswang1029** [comment]
supermosaic.10.fits, 100% clip
clip min = -96.74206926673651
clip max = 92.02825746685266

supermosaic.10.hdf5, 100 clip
clip min = -96.74207305908203
clip max = 92.0282249301672

it seems there is difference.

**Comment by kswang1029** [comment]
hdf5
histograms: Array(1)
0: e
binWidth: 0.06797634065151215
bins: (2776) [1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, …]
firstBinCenter: -96.70808488875628
numBins: 2776

fits:
binWidth: 0.06800083816051483
bins: (2775) [1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, …]
firstBinCenter: -96.70806884765625
numBins: 2775

**Comment by kswang1029** [comment]
The numBins are different by 1 and I guess it comes from rounding sqrt(npix_x * npix_y).
I think this difference won't affect users much since the intention of this histogram is for clipping.

@arusstaylor @veggiesaurus do you think we need to make them consistent? (either fix hdf5 converter or fix the carta backend)

**Comment by veggiesaurus** [comment]
@kswang1029 nice find. I used `ceil(sqrt(nx * ny))` to avoid having zero bins. Probably easier to change CARTA, rather than changing the converter and re-converting all HDF5 files.

**Comment by arusstaylor** [comment]
so problem solved?

**Comment by veggiesaurus** [comment]
> so problem solved?

I think the solution is quite simple (adjust the rounding in CARTA). But that might require the tests to be adjusted, @kswang1029

**Comment by kswang1029** [comment]
Yes, if we decided to modify carta backend, I can adjust the relevant tests accordingly. We just need a decision on if we want to make them consistent and when to do that. :)

**Comment by veggiesaurus** [comment]
@confluence assigning you this minor issue for when you have time.

**Comment by confluence** [comment]
It looks like the Python converter uses the same method as CARTA (I have to check to make 100% sure that the result is the same), so it's possible that the two converters have already been producing different files.

We're about to be regenerating files anyway because we're changing the schema to store `SUM` and `SUM_SQ` statistics instead of `MEAN`. So, practical considerations aside, which solution is objectively preferable?

If I don't hear any objections, I'm going to standardise to the way the C converter does it.

**Comment by kswang1029** [comment]
@confluence if there is a new hdf5 convertor, please let me know. We will need to regenerate the hdf5 images in our CARTA image pool as well.

**Comment by confluence** [comment]
@kswang1029 We'll definitely let you know -- before we can start using that version we will need to make a change to the backend to use the new stat types, so it isn't going to happen immediately.

**Comment by confluence** [comment]
This is fixed in the c++ converter in the development branch, which we will be merging soon. I will later also update the Python converter.

**Comment by confluence** [comment]
@kswang1029 Once you have generated some new files using the latest converter version, please confirm whether this is fixed.

**Comment by kswang1029** [comment]
@confluence could you provide a guide line about how to build the convertor (dev branch?)?

**Comment by ajm-asiaa** [comment]
@confluence Our systems here run RedHat7. The problem is that the hdf5_serial library doesn't exist for RedHat7. I remember we had a similar problem when we first tried to compile the original hdf5-image-viewer on RedHat, but I can't find any discussion or documentation on how it was solved. 
It seems it needs more than simply changing 
>set(LINK_LIBS ${LINK_LIBS} cfitsio hdf5_serial hdf5_cpp hdf5_hl_cpp) 

to 

> set(LINK_LIBS ${LINK_LIBS} cfitsio ${HDF5_LIBRARIES}) 

plus adding things like 
>FIND_PACKAGE(HDF5)
>INCLUDE_DIRECTORIES(${HDF5_INCLUDE_DIR})
> add_definitions(-DHAVE_HDF5) 

to the CMakeLists.txt file
